baseapp/baseapp.go

```diff
@@ -761,3 +768,106 @@ func (app *BaseApp) runMsgs(ctx sdk.Context, msgs []sdk.Msg, mode runTxMode) (*s
+	storeToLoadFrom := make(map[string]types.KVStore)
+	storeKeyNames := make([]string, 0, len(storeKeys))
+	for _, storeKey := range storeKeys {
+		storeKeyName := storeKey.Name()
+		storeKeyNames = append(storeKeyNames, storeKeyName)
+		storeToLoadFrom[storeKeyName] = previousCMS.GetKVStore(storeKey)
+		storeKeyToSubstoreTraceBuf[storeKey] = &bytes.Buffer{}
+	}
+
+	// BaseApp, B1
+	options := []AppOption{
+		SetSubstoreTracer(storeTraceBuf),
+		SetTracerFor(storeKeys[0].Name(), storeKeyToSubstoreTraceBuf[storeKeys[0]]),
+	}
+	newApp, err := SetupBaseAppFromParams(app.name, app.logger, dbm.NewMemDB(), app.txDecoder, storeKeyNames, app.router, app.LastBlockHeight()+1, storeToLoadFrom, options...)
+
+	return newApp, storeKeyToSubstoreTraceBuf, err
+}
+
+func (app *BaseApp) generateFraudProof(storeKeyToSubstoreTraceBuf map[types.StoreKey]*bytes.Buffer) (FraudProof, error) {
+	fraudProof := FraudProof{}
+	fraudProof.stateWitness = make(map[string]StateWitness)
+	cms := app.cms.(*multi.Store)
+
+	for storeKey, subStoreTraceBuf := range storeKeyToSubstoreTraceBuf {
+		keys := cms.GetKVStore(storeKey).(*tracekv.Store).GetAllKeysUsedInTrace(*subStoreTraceBuf)
+
+		substoreSMT := cms.GetSubstoreSMT(storeKey.Name())
+		proof, storeHash, err := cms.GetSubStoreProof(storeKey.Name())
+		if err != nil {
+			return FraudProof{}, err
+		}
+		stateWitness := StateWitness{
+			proof:       *proof,
+			rootHash:    storeHash,
+			WitnessData: make([]WitnessData, 0, keys.Len()),
+		}
+		for key := range keys {
+			value := substoreSMT.Get([]byte(key))
+			proofOps, err := substoreSMT.GetProof([]byte(key))
+			if err != nil {
+				return FraudProof{}, err
+			}
+			proofOp := proofOps.GetOps()[0]
+			bKey, bVal := []byte(key), []byte(value)
+			witnessData := WitnessData{bKey, bVal, proofOp}
+			stateWitness.WitnessData = append(stateWitness.WitnessData, witnessData)
+		}
+		fraudProof.stateWitness[storeKey.Name()] = stateWitness
+	}
+
+	return fraudProof, nil
+}
+
+func SetupBaseAppFromParams(appName string, logger log.Logger, db dbm.Connection, txDecoder sdk.TxDecoder, storeKeyNames []string, router sdk.Router, blockHeight int64, storeToLoadFrom map[string]types.KVStore, options ...AppOption) (*BaseApp, error) {
+	storeKeys := make([]types.StoreKey, 0, len(storeKeyNames))
+	for _, storeKeyName := range storeKeyNames {
+		storeKey := sdk.NewKVStoreKey(storeKeyName)
+		storeKeys = append(storeKeys, storeKey)
+		subStore := storeToLoadFrom[storeKeyName]
+		it := subStore.Iterator(nil, nil)
+		for ; it.Valid(); it.Next() {
+			key, val := it.Key(), it.Value()
+			options = append(options, SetSubstoreKVPair(storeKey, key, val))
+		}
+	}
+	options = append(options, SetSubstores(storeKeys...))
+
+	// This initial height is used in `BeginBlock` in `validateHeight`
+	options = append(options, SetInitialHeight(blockHeight))
+
+	// make list of options to pass by parsing fraudproof
+	app := NewBaseApp(appName, logger, db, txDecoder, options...)
+
+	app.router = router
```

Comment:
what’s the best way to reuse a route from a one app B1 to another app B2? Just passing in the same router from B1 doesn’t seem to work  because during the route it uses B1 but I want it to use B2.